drivers/hmis_external_apis/app/models/hmis_external_apis/ac_hmis/importers/projects_importer.rb

```diff
@@ -0,0 +1,222 @@
+    end
+
+    def validate
+      Rails.logger.info 'Validating CSVs (wip)'
+    end
+
+    def upsert_funders
+      generic_upsert(
+        file: 'Funder.csv',
+        conflict_target: ['"FunderID"', 'data_source_id'],
+        klass: GrdaWarehouse::Hud::Funder,
+      )
+    end
+
+    def upsert_orgs
+      generic_upsert(
+        file: 'Organization.csv',
+        conflict_target: ['"OrganizationID"', 'data_source_id'],
+        klass: GrdaWarehouse::Hud::Organization,
+      )
+    end
+
+    def upsert_projects
+      @project_result = generic_upsert(
+        file: 'Project.csv',
+        conflict_target: ['"ProjectID"', 'data_source_id'],
+        klass: GrdaWarehouse::Hud::Project,
+        ignore_columns: ['Walkin'],
+      )
+    end
+
+    # FIXME: I'm not sure how to execute this efficiently since I think we might have to fetch the user from the project.
+    # FIXME: What do we do if there's no user? I'm just skipping them for now.
+    def upsert_walkins
+      project_ids = @project_result.ids
+      walkin = records_from_csv('Project.csv').map { |x| x['Walkin'] }
+
+      project_ids.length != walkin.length and raise(AbortImportException, 'Project upsert should have been the same length as the parsed csv')
+
+      project_ids.zip(walkin).each do |(project_id, bool_str)|
+        project = Hmis::Hud::Project.find(project_id)
+
+        if project.user.blank?
+          Rails.logger.error "Can't create a custom data element without a user, and the project doesn't have a user for Project #{project_id}"
+          next
+        end
+
+        cde = Hmis::Hud::CustomDataElement
+          .where(
+            owner: project,
+            # FIXME: we could do this to make this all faster, but we need the user I think.
+            # owner_type: 'Hmis::Hud::Project',
+            # owner_id: project_id,
+            data_element_definition: cded,
+            data_source: data_source,
+            user: project.user,
+          )
+          .first_or_initialize
+
+        cde.value_boolean =
+          case bool_str
+          when 'Y' then true
+          when 'N' then false
+          end
+
+        cde.save!
+      end
+    end
+
+    def analyze
+      ProjectsImportAttempt.connection.exec_query('ANALYZE VERBOSE "Funder", "Project", "Organization", "CustomDataElements";')
+    end
+
+    def finish
+      attempt.status = ProjectsImportAttempt::SUCCEEDED
+      attempt.save!
+    end
+
+    def records_from_csv(file)
+      io = File.open(file, 'r')
+
+      # Checking for BOM
+      if io.read(3).bytes == [239, 187, 191]
+        Rails.logger.info 'Byte-order marker (BOM) found. Skipping it.'
+      else
+        io.rewind
+      end
+
+      # FIXME: Are these going to be huge?
+      CSV.parse(io.read, headers: true, skip_lines: /\A\s*\z/)
+    end
+
+    # FIXME: This feels like something that might already exist, but I didn't
+    # find a good candidate that I felt safe modifying or reusing. Let me know if it exists
```

Comment:
I don't think we have anything exactly like this.  We tend to load the data from the CSV, do some pre-processing and then the import.  For the main importer, we had to speed things up significantly, so we drop into native postgres CSV loading see `HmisCsvImporter::Loader::Loader#load_source_file_pg`.  But that gets loaded into an all string-table, then converted to the right format along with other data cleanup.

`Health::Tasks::ImportEpic#import` is also similar, but comes off SFTP.  You might want to crib the CSV reading from there if you run into issues with real files.  We've seen BOMs all over the map.

Reply:
Do you want me to load into a staging table instead of directly into the live tables?